File: packages/flutter-im-ui/test/moments_privacy_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter_test/flutter_test.dart';

// FR-100: moments privacy in words, not SDK codes.

void main() {
  test('the kit enums are the shared vocabulary', () {
    final vocabulary =
        jsonDecode(
              File(
                '${Directory.current.path}/../../spec/moments-privacy.json',
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    expect(
      FlareMomentVisibility.values.map((v) => v.name).toList(),
      (vocabulary['visibility'] as List).cast<String>(),
    );
    expect(
      FlareMomentAudienceMode.values.map((v) => v.name).toList(),
      (vocabulary['audienceMode'] as List).cast<String>(),
    );
    expect(
      FlareMomentHistoryRange.values.map((v) => v.name).toList(),
      (vocabulary['historyRange'] as List).cast<String>(),
    );
    expect(
      FlareMomentVisibility.values.where(flareMomentAudienceApplies).map((v) => v.name).toList(),
      ['friends', 'public'],
    );
  });
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/MomentsPrivacyTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals

/** FR-100: moments privacy in words, not SDK codes. */
class MomentsPrivacyTest {
    private fun vocabularyFile(): File {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "spec/moments-privacy.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile
        }
        error("spec/moments-privacy.json not found")
    }

    private fun names(values: List<Enum<*>>) = values.map { it.name.replaceFirstChar { c -> c.lowercase() } }

    @Test fun theKitEnumsAreTheSharedVocabulary() {
        val vocabulary = FlareJson.parse(vocabularyFile().readText()) as Map<*, *>
        assertEquals(vocabulary["visibility"], names(FlareMomentVisibility.entries))
        assertEquals(vocabulary["audienceMode"], names(FlareMomentAudienceMode.entries))
        assertEquals(vocabulary["historyRange"], names(FlareMomentHistoryRange.entries))
        assertEquals(
            listOf("friends", "public"),
            names(FlareMomentVisibility.entries.filter { flareMomentAudienceApplies(it) }),
        )
    }
}
